### PR TITLE
feat(FRM): add missing fields in Signifyd payment request

### PR DESCRIPTION
**Pull request by srujanchikke**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [x] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Following fields are added in siginfyd sale/transaction request :

order_id, subcategory, category, confirmation_phone_number , confirmation email, shipping method, total shipping cost 

### Additional Changes

- [X] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

```
"order_details" : [{
        "product_name" : "gillete creme",
        "quantity" : 2,
        "amount" : 600,
        "product_id" : "product_id",
        "category" : "clothes", 
        "sub_category" : "tees" // newly added
    }]
```

## How did you test it?

Postman 
Test cases : configure for both pre and post frm flows and make payment 
```
{
  "amount": 620,
  "currency": "USD",
  "confirm": true,
  "capture_method": "automatic",
  "capture_on": "2022-09-10T10:11:12Z",
  "amount_to_capture": 620,
  "customer_id": "StripeCustomer2",
  "name": "Bob Smith",
  "phone": "999999999",
  "phone_country_code": "+91",
  "description": "Its my first payment request",
  "authentication_type": "no_three_ds",
  "return_url": "https://google.com",
  "payment_method": "card",
  "payment_method_data": {
    "card": {
      "card_number": "4242424242424242",
      "card_exp_month": "10",
      "card_exp_year": "25",
      "card_holder_name": "joseph Doe",
      "card_cvc": "123"
    }
  },
  "billing": {
    "address": {
      "line1": "1467",
      "line2": "Harrison Street",
      "line3": "Harrison Street",
      "city": "San Fransico",
      "state": "California",
      "zip": "94122",
      "country": "US",
      "first_name": "joseph",
      "last_name": "Doe"
    },
    "phone": {
      "number": "8056594427",
      "country_code": "+91"
    }
  },
  "shipping": {
    "address": {
      "line1": "1467",
      "line2": "31 Sherwood Gardens",
      "line3": "31 Sherwood Gardens",
      "city": "London",
      "state": "Manchester",
      "zip": "E14 9wn",
      "country": "GB",
      "first_name": "Bob",
      "last_name": "Smith"
    },
    "phone": {
      "number": "8056594427",
      "country_code": "+91"
    }
  },
  "statement_descriptor_name": "joseph",
  "statement_descriptor_suffix": "JS",
  "email" : "bob@gmail.com",
  "metadata": {
    "udf1": "value1",
    "new_customer": "true",
    "login_date": "2019-09-10T10:11:12Z"
  },
  "order_details" : [{
        "product_name" : "gillete creme",
        "quantity" : 2,
        "amount" : 600,
        "product_id" : "{{random_number}}",
        "category" : "clothes",
        "sub_category" : "tees",
        "product_type" : "digital"
    },
    {
        "product_name" : "gillete razor",
        "quantity" : 1,
        "amount" : 300,
        "product_id" : "{{random_number}}",
        "category" : "clothes",
        "sub_category" : "tees",
        "product_type" : "digital"
    }],
    "frm_metadata" : {
        "fulfillment_method" : "delivery",
        "coverage_request" : "fraud",
        "total_shipping_cost" : 240,
        "order_channel" : "web"
    }
}
```
fulfillment API : 
```
{
    "payment_id": "{{payment_id}}",
    "order_id": "{{payment_id}}_1",
    "fulfillments": [
        {
            "shipment_id": "{{$randomUUID}}",
            "destination": {
                "full_name": "Bob Smith",
                "organization": "Initrode plc",
                "email": null,
                "confirmation_phone": "+918900384478",
                "address": {
                    "street_address": "31 Sherwood Gardens",
                    "unit": "string",
                    "postal_code": "E14 9wn",
                    "city": "London",
                    "province_code": "string",
                    "country_code": "GB"
                }
            }
        }
    ],
    "tracking_urls": [
        "string"
    ],
    "tracking_numbers": [
        "string"
    ],
    "fulfillment_method": "DELIVERY",
    "carrier": "UPS",
    "shipment_status": "WAITING_FOR_PICKUP",
    "shipped_at": "2019-08-24T14:15:22Z"
}
```

- verify it from signifyd dashboard sale/checkout/transaction api's

Screenshot 2024-05-06 at 4 03 18 PM
Screenshot 2024-05-06 at 4 03 33 PM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

**Review by deepanshu-iiitu — DISMISSED** [review]

File: crates/router/src/connector/signifyd/transformers/api.rs [hunk 1]
```diff
@@ -84,12 +107,31 @@ pub struct Address {
     country_code: common_enums::CountryAlpha2,
 }
 
+#[derive(Debug, Serialize, Eq, PartialEq, Deserialize, Clone)]
+#[serde(rename_all(serialize = "SCREAMING_SNAKE_CASE", deserialize = "snake_case"))]
+pub enum CoverageRequests {
+    Fraud,
+    Inr,
```
Comment (sai-harsha-vardhan) [inline crates/router/src/connector/signifyd/transformers/api.rs:114]: Give proper renaming and use serde rename here

File: crates/router/src/connector/signifyd/transformers/api.rs [hunk 2]
```diff
@@ -103,9 +145,25 @@ impl TryFrom<&frm_types::FrmSaleRouterData> for SignifydPaymentsSaleRequest {
                 item_name: order_detail.product_name.clone(),
                 item_price: order_detail.amount,
                 item_quantity: i32::from(order_detail.quantity),
+                item_id: order_detail.product_id.clone(),
+                item_category: order_detail.category.clone(),
+                item_sub_category: order_detail.sub_category.clone(),
+                item_is_digital: Some(
```
Comment (sai-harsha-vardhan) [inline crates/router/src/connector/signifyd/transformers/api.rs:151]: Use `map` here

File: crates/router/src/connector/signifyd/transformers/api.rs [hunk 3]
```diff
@@ -128,19 +186,34 @@ impl TryFrom<&frm_types::FrmSaleRouterData> for SignifydPaymentsSaleRequest {
         };
 
         let created_at = common_utils::date_time::now();
-        let order_channel = OrderChannel::Web;
-        let shipments = Shipments { destination };
+        let order_channel: OrderChannel = OrderChannel::Web;
```
Comment (sai-harsha-vardhan) [inline crates/router/src/connector/signifyd/transformers/api.rs:189]: Why's this hardcoded to Web?

File: crates/router/src/connector/signifyd/transformers/api.rs [hunk 4]
```diff
@@ -128,19 +186,34 @@ impl TryFrom<&frm_types::FrmSaleRouterData> for SignifydPaymentsSaleRequest {
...
+            confirmation_email: item.request.email.clone(),
+            confirmation_phone: billing_address
+                .clone()
+                .phone
+                .and_then(|phone_data| phone_data.number),
         };
         Ok(Self {
             order_id: item.attempt_id.clone(),
             purchase,
             decision_delivery: DecisionDelivery::Sync,
```
Comment (sai-harsha-vardhan) [inline crates/router/src/connector/signifyd/transformers/api.rs:213]: Please provide the comment for reason of this hardcoding

File: crates/router/src/core/fraud_check/flows/sale_flow.rs [hunk 5]
```diff
@@ -65,6 +66,12 @@ impl ConstructFlowSpecificData<frm_api::Sale, FraudCheckSaleData, FraudCheckResp
             request: FraudCheckSaleData {
                 amount: self.payment_attempt.amount,
                 order_details: self.order_details.clone(),
+                currency: self.payment_attempt.currency,
+                email: customer.clone().and_then(|customer_data| {
+                    customer_data
+                        .email
+                        .and_then(|email| Email::try_from(email.into_inner().expose()).ok())
```
Comment (sai-harsha-vardhan) [inline crates/router/src/core/fraud_check/flows/sale_flow.rs:73]: Shouldn't we propagate the error here?

File: crates/router/src/connector/signifyd/transformers/api.rs (hunk 3 above)
Comment (srujanchikke) [reply crates/router/src/connector/signifyd/transformers/api.rs]: This was hardcoded before, moving this field to payment frm metadata

File: crates/router/src/core/fraud_check/flows/sale_flow.rs (hunk 5 above)
Comment (srujanchikke) [reply crates/router/src/core/fraud_check/flows/sale_flow.rs]: Yes, updated!

**Review by deepanshu-iiitu — APPROVED** [review]

**Review by sai-harsha-vardhan — APPROVED** [review]

**Pull request merged**